HackAtUCI/irvinehacks-site
# Restructure user record to have multiple roles

Potentially controversial change to how user roles are managed. Expecting this will allow for more flexibility for new admin features and also given the need to support other types of applicants in #386.

Resolves #421. See description of changes for notable modifications besides renaming the field.

#### Changes
- With the end goal of having more flexibility in assigning user roles, make `UserRecord` store multiple possible roles
  - Field is renamed to plural and modeled as a tuple or set
- Provide empty array in default user identity instead of `null`
  - Semantically `null` could be better to indicate unknown rather than explicitly no roles, but functionally, an empty array is simpler
- Constrain database query for `check_in_participant`
  - Still need unit tests for this
- Update role and status check for `confirm_attendance_non_hacker`
- Use `AfterValidator` for `Applicant.roles` to include `Role.APPLICANT`
- Reorganize admin authorization roles to reuse dependencies
  - Basic admin access is provided to anybody with the "Organizer" role
  - Directors will need to have both "Organizer" and "Director" roles
- Update logic for checking if user has already applied
- Fix logical bug in `ParticipantAction`: function was used like boolean
- Rename role strings to be the same as their display labels

#### Testing
Note any older records in the local database will need to be updated.
1. Impersonate a new user and submit an application
2. Observe in the database that the applicant has the correct default role: array with single entry of "Applicant"
3. Add another sample user record with the "Organizer" role
4. Impersonate said user and confirm the admin dashboard is accessible
5. Add an application manager role, e.g. "Reviewer" to said user
6. Observe the Applicants page of the admin site is accessible and the previously submitted application is included
